### PR TITLE
docker_watcher: Move docker_watcher into pkg/workloads/containerd

### PR DESCRIPTION
**Pull request by tgraf**
* Unties the docker watcher from the daemon and isolates it in a new
  workloads package. This prepares to support other runtimes in the
  future.

* Unties IPAM from the daemon and moves it into `pkg/ipam`. Only the
  the API handler layer remains in the daemon.

Fixes: #1595

Signed-off-by: Thomas Graf <thomas@cilium.io>

### CODE REVIEW

File: daemon/config.go [hunk 1]
```diff
@@ -43,22 +42,18 @@ const (
...
 // Config is the configuration used by Daemon.
 type Config struct {
-	BpfDir         string     // BPF template files directory
-	LibDir         string     // Cilium library files directory
-	RunDir         string     // Cilium runtime directory
-	NAT46Prefix    *net.IPNet // NAT46 IPv6 Prefix
-	Device         string     // Receive device
-	HostV4Addr     net.IP     // Host v4 address of the snooping device
-	HostV6Addr     net.IP     // Host v6 address of the snooping device
-	DockerEndpoint string     // Docker endpoint
```
Comment (aanm) [inline daemon/config.go:53]: What was `DockerEndpoint` used for?

File: daemon/endpoint.go [hunk 2]
```diff
@@ -647,3 +609,65 @@ func (h *putEndpointIDLabels) Handle(params PutEndpointIDLabelsParams) middlewar
...
+	oldIdentity := ep.GetIdentity()
+	ep.SetIdentity(d, identity)
+	ep.Mutex.Unlock()
+
+	// Skip building endpoint if identity is invalid or unchanged
+	if identity.ID != oldIdentity {
+		ep.Regenerate(d)
+
+		// FIXME: Does this rebuild epID twice?
+		d.TriggerPolicyUpdates([]policy.NumericIdentity{identity.ID})
```
Comment (ianvernon) [inline daemon/endpoint.go:670]: `d.TriggerPolicyUpdates` triggers endpoint regeneration (see code snippet below from daemon's `TriggerPolicyUpdates`), so I don't think you need `ep.Regenerate(d)` on line 666:
```
for k := range Endpoints {
		go func(ep *endpoint.Endpoint, wg *sync.WaitGroup) {
			policyChanges, err := ep.TriggerPolicyUpdates(owner)
			if err != nil {
				log.Warningf("Error while handling policy updates for endpoint %s", err)
				ep.LogStatus(endpoint.Policy, endpoint.Failure, err.Error())
			} else {
				ep.LogStatusOK(endpoint.Policy, "Policy regenerated")
			}
			if policyChanges {
				<-ep.Regenerate(owner)
			}
			wg.Done()
		}(Endpoints[k], &wg)
	}
```

File: daemon/state.go [hunk 3]
```diff
@@ -23,14 +23,28 @@ import (
...
+	"github.com/cilium/cilium/pkg/logfields"
 	"github.com/cilium/cilium/pkg/policy"
+	"github.com/cilium/cilium/pkg/workloads/containerd"
 
-	dockerAPI "github.com/docker/engine-api/client"
 	log "github.com/sirupsen/logrus"
-	ctx "golang.org/x/net/context"
 )
 
+// cleanUpDockerDanglingEndpoints cleans all endpoints that are dandling by
```
Comment (ianvernon) [inline daemon/state.go]: dangling

File: daemon/main.go [hunk 4]
```diff
@@ -90,6 +91,7 @@ var (
 	v6ServicePrefix       string
 	k8sAPIServer          string
 	k8sKubeConfigPath     string
+	dockerEndpoint        string
```
Comment (aanm) [inline daemon/main.go:94]: Don't you want to move this to `pkg/containerd`?

File: daemon/state_test.go [hunk 5]
```diff
@@ -180,85 +172,10 @@ func (ds *DaemonSuite) TestReadEPsFromDirNames(c *C) {
...
-// Helper function to mock docker calls
-type transportFunc func(*http.Request) (*http.Response, error)
-
-// Helper function to mock docker calls
-func (tf transportFunc) RoundTrip(req *http.Request) (*http.Response, error) {
-	return tf(req)
-}
-
-// Helper function to mock docker calls
-func newMockClient(doer func(*http.Request) (*http.Response, error)) *http.Client {
```
Comment (aanm) [inline daemon/state_test.go]: Where's this?

File: pkg/ipam/ipam_test.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+func Test(t *testing.T) {
+	TestingT(t)
+}
+
+type IPAMSuite struct{}
+
+var _ = Suite(&IPAMSuite{})
+
+func (s *IPAMSuite) TestLock(c *C) {
+	Init()
```
Comment (aanm) [inline pkg/ipam/ipam_test.go]:
```
err := Init()
c.Assert(err, IsNil)
```

File: pkg/ipam/types.go [hunk 7]
```diff
@@ -12,24 +12,19 @@
...
-	// FIXME should we calculate LabelsHash here?
-	return &Container{
-		ContainerJSON: *dc,
-	}
+// Config is the IPAM configuration used for a particular IPAM type.
+type Config struct {
+	IPAMConfig     allocator.IPAMConfig
+	IPv6Allocator  *ipallocator.Range
+	IPv4Allocator  *ipallocator.Range
+	AllocatorMutex lock.RWMutex
```
Comment (aanm) [inline pkg/ipam/types.go]: - Can't the mutex be private?

- Can you put the allocator in the line(s) above the parameters that is protecting?

File: pkg/workloads/containerd/client.go [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+		return err
+	}
+
+	dockerClient = c
+
+	return nil
+}
+
+// InitMock initializes the containerd packged with a mocked runtime
+func InitMock() error {
```
Comment (aanm) [inline pkg/workloads/containerd/client.go:55]: Does this (L65-L138) needs to be in here? Can't this be in `_test.go`?

**Comment by ianvernon** [comment]
Overall, this looks fantastic! The more code we move out of the daemon, the better.

File: pkg/workloads/containerd/watcher.go [hunk 9]
```diff
@@ -49,36 +50,63 @@ const (
...
 
-// SyncDocker is used by the daemon to synchronize changes between Docker and
+// syncWithRuntime is used by the daemon to synchronize changes between Docker and
 // Cilium. This includes identities, labels, etc.
-func (d *Daemon) SyncDocker() {
+func syncWithRuntime() {
 	var wg sync.WaitGroup
 
-	cList, err := d.dockerClient.ContainerList(ctx.Background(), dTypes.ContainerListOptions{All: false})
+	// FIXME: Must be synchronize with event handler
```
Comment (aanm) [inline pkg/workloads/containerd/watcher.go]: Can you clarify this a little more?

**Review by aanm — CHANGES_REQUESTED** [review]
So we can't support 2 container runtimes at the same time? I'm asking based on the overall code + the cilium status function where we are only checking for containerd status.

File: daemon/endpoint.go (hunk 2 above)
Comment (aanm) [reply daemon/endpoint.go:670]: I think we should tackle on which occasions we need an endpoint regeneration and on which occasions we are regenerating unnecessary. For now are regenerating away more than we need to but it's something we clearly need to fix.

Comment (tgraf) [reply daemon/endpoint.go:670]: I agree with Ian, this PR is leaving this code unmodified. I just moved it over. I'd rather tackle generation optimization in a separate PR.

File: daemon/state.go (hunk 3 above)
Comment (tgraf) [reply daemon/state.go]: fixed

**Comment by tgraf** [comment]
> So we can't support 2 container runtimes at the same time? I'm asking based on the overall code + the cilium status function where we are only checking for containerd status.

This PR does not address that but eventually, we should add that.

File: daemon/config.go (hunk 1 above)
Comment (tgraf) [reply daemon/config.go:53]: This is now passed in as via `func Init(endpoint string) error {`

File: daemon/main.go (hunk 4 above)
Comment (tgraf) [reply daemon/main.go:94]: This is just a temporary storage place to pass it into the `Init()` function later.

File: pkg/ipam/ipam_test.go (hunk 6 above)
Comment (tgraf) [reply pkg/ipam/ipam_test.go]: fixed

File: pkg/ipam/types.go (hunk 7 above)
Comment (tgraf) [reply pkg/ipam/types.go]: made it private. the mutex protects the entire struct. added a comment.

File: pkg/workloads/containerd/watcher.go (hunk 9 above)
Comment (tgraf) [reply pkg/workloads/containerd/watcher.go]: docker events and the manual sync can overlap and we could readd a container after we have removed it based on the event if the list occurred before the delete event was sent out. I'll file a github issue instead.

File: pkg/workloads/containerd/client.go (hunk 8 above)
Comment (tgraf) [reply pkg/workloads/containerd/client.go:55]: I don't mind it being in `client.go`. Keeps the related code together.

**Review by aanm — APPROVED** [review]

File: daemon/endpoint.go (hunk 2 above)
Comment (ianvernon) [reply daemon/endpoint.go:670]: Ah, OK - we can address this later, then.

**Review by ianvernon — APPROVED** [review]

File: pkg/workloads/containerd/client.go (hunk 8 above)
Comment (aanm) [reply pkg/workloads/containerd/client.go:55]: But you could have the `client_test.go`

Comment (tgraf) [reply pkg/workloads/containerd/client.go:55]: How? The code will no longer be included.

Comment (aanm) [reply pkg/workloads/containerd/client.go:55]: I thought InitMock was not being used outside the `containerd` pkg

**Pull request merged**